# Basic Python Syntax

In [33]:
# set the midpoint
midpoint = 5

# make two empty lists
lower = []; upper = []

# split the numbers into lower and upper
for i in range(10):
    if (i < midpoint):
        lower.append(i)
    else:
        upper.append(i)
        
print("lower:", lower)
print("upper:", upper)


lower: [0, 1, 2, 3, 4]
upper: [5, 6, 7, 8, 9]


In [35]:
x=1+2
x = 1 + 2
x             =        1    +                2

In [37]:
2 * (3 + 4)

14

In [38]:
print('first value:', 1)

first value: 1


In [39]:
print('second value:', 2)

second value: 2


In [40]:
L = [4,2,3,1]
L.sort()
print(L)

[1, 2, 3, 4]


# Variables and Objects

In [41]:
x = 1         # x is an integer
x = 'hello'   # now x is a string
x = [1, 2, 3] # now x is a list

In [43]:
x = [1, 2, 3]
y = x

In [44]:
print(y)

[1, 2, 3]


In [45]:
x.append(4) # append 4 to the list pointed to by x
print(y) # y's list is modified as well!

[1, 2, 3, 4]


In [46]:
x = "something else"
print(y)

[1, 2, 3, 4]


In [47]:
x = 10
y = x
x += 5  # add 5 to x's value, and assign it to x
print("x =", x)
print("y =", y)

x = 15
y = 10


In [48]:
x = 4
type(x)

int

In [49]:
x = 'hello'
type(x)

str

In [50]:
x = 3.14159
type(x)

float

In [51]:
L = [1, 2, 3]
L.append(100)
print(L)

[1, 2, 3, 100]


In [52]:
x = 4.5
print(x.real, "+", x.imag, 'i')

4.5 + 0.0 i


In [53]:
x = 4.5
x.is_integer()

False

In [54]:
x = 4.0
x.is_integer()

True

In [55]:
type(x.is_integer)

builtin_function_or_method

# Operators

**Operator	Name	Description**

a + b	Addition	Sum of a and b

a - b	Subtraction	Difference of a and b

a * b	Multiplication	Product of a and b

a / b	True division	Quotient of a and b

a // b	Floor division	Quotient of a and b, removing fractional parts

a % b	Modulus	Integer remainder after division of a by b

a ** b	Exponentiation	a raised to the power of b

-a	Negation	The negative of a

+a	Unary plus	a unchanged (rarely used)

**Operator	Name	Description**

a & b	Bitwise AND	Bits defined in both a and b

a | b	Bitwise OR	Bits defined in a or b or both

a ^ b	Bitwise XOR	Bits defined in a or b but not both

a << b	Bit shift left	Shift bits of a left by b units

a >> b	Bit shift right	Shift bits of a right by b units

~a	Bitwise NOT	Bitwise negation of a

a += b	a -= b	a *= b	a /= b

a //= b	a %= b	a **= b	a &= b

a |= b	a ^= b	a <<= b	a >>= b

**Operation	Description		Operation	Description**

a == b	a equal to b		a != b	a not equal to b

a < b	a less than b		a > b	a greater than b

a <= b	a less than or equal to b		a >= b	a greater than or equal to b

**Operator	Description**

a is b	True if a and b are identical objects

a is not b	True if a and b are not identical objects

a in b	True if a is a member of b

a not in b	True if a is not a member of b

In [56]:
# addition, subtraction, multiplication
(4 + 8) * (6.5 - 3)

42.0

In [57]:
# True division
print(11 / 2)

5.5


In [58]:
# Floor division
print(11 // 2)

5


In [59]:
bin(10)

'0b1010'

In [60]:
bin(4)

'0b100'

In [61]:
4 | 10

14

In [62]:
bin(4 | 10)

'0b1110'

In [63]:
a = 24
print(a)

24


In [64]:
a + 2

26

In [65]:
a += 2  # equivalent to a = a + 2
print(a)

26


In [66]:
# 25 is odd
25 % 2 == 1

True

In [67]:
# 66 is odd
66 % 2 == 1

False

In [68]:
# check if a is between 15 and 30
a = 25
15 < a < 30

True

In [69]:
#flipping bits
-1 == ~0

True

In [70]:
x = 4
(x < 6) and (x > 2)

True

In [71]:
(x > 10) or (x % 2 == 0)

True

In [72]:
not (x < 6)

False

In [73]:
# (x > 1) xor (x < 10)
(x > 1) != (x < 10)

False

In [74]:
a = [1, 2, 3]
b = [1, 2, 3]

In [75]:
a == b

True

In [76]:
a is b

False

In [77]:
a is not b

True

In [78]:
a = [1, 2, 3]
b = a
a is b

True

In [79]:
1 in [1, 2, 3]

True

In [80]:
2 not in [1, 2, 3]

False

# Simple Values

In [81]:
x = 1
type(x)

int

In [82]:
2 ** 200

1606938044258990275541962092341162602522202993782792835301376

In [83]:
5 / 2

2.5

In [84]:
5 // 2

2

In [85]:
x = 0.000005
y = 5e-6
print(x == y)

True


In [86]:
x = 1400000.00
y = 1.4e6
print(x == y)

True


In [87]:
float(1)

1.0

In [88]:
0.1 + 0.2 == 0.3

False

In [89]:
print("0.1 = {0:.17f}".format(0.1))
print("0.2 = {0:.17f}".format(0.2))
print("0.3 = {0:.17f}".format(0.3))

0.1 = 0.10000000000000001
0.2 = 0.20000000000000001
0.3 = 0.29999999999999999


In [90]:
complex(1, 2)

(1+2j)

In [91]:
1 + 2j

(1+2j)

In [92]:
c = 3 + 4j

In [93]:
c.real  # real part

3.0

In [94]:
c.imag  # imaginary part

4.0

In [95]:
c.conjugate()  # complex conjugate

(3-4j)

In [96]:
abs(c)  # magnitude, i.e. sqrt(c.real ** 2 + c.imag ** 2)

5.0

In [97]:
message = "what do you like?"
response = 'spam'

In [98]:
# length of string
len(response)

4

In [99]:
# Make upper-case. See also str.lower()
response.upper()

'SPAM'

In [100]:
# Capitalize. See also str.title()
message.capitalize()

'What do you like?'

In [101]:
# concatenation with +
message + response

'what do you like?spam'

In [102]:
# multiplication is multiple concatenation
5 * response

'spamspamspamspamspam'

In [103]:
# Access individual characters (zero-based indexing)
message[0]

'w'

In [104]:
type(None)

NoneType

In [105]:
return_value = print('abc')

abc


In [106]:
print(return_value)

None


In [107]:
result = (4 < 5)
result

True

In [108]:
type(result)

bool

In [109]:
print(True, False)

True False


In [110]:
#bool() on numeric type is Flase if == 0, True otherwise
bool(2014)

True

In [111]:
bool(0)

False

In [112]:
bool(3.1415)

True

In [113]:
#bool()type of none is always False
bool(None)

False

In [114]:
#bool() on empty strings are always False, True for else|
bool("")

False

In [115]:
bool("abc")

True

# Built-In Data Structures

In [118]:
#for sequences, empty = False, else is true
bool([1, 2, 3])

True

In [119]:
bool([])

False

In [32]:
L = [1, 'two', 3.14, [0, 3, 5]]
print(L)

[1, 'two', 3.14, [0, 3, 5]]


In [2]:
L[0] = 100
print (L)

[100, 'two', 3.14, [0, 3, 5]]


In [3]:
L

[100, 'two', 3.14, [0, 3, 5]]

In [4]:
L[1:3] = [55, 56]

In [5]:
L

[100, 55, 56, [0, 3, 5]]

In [6]:
t = (1, 2, 3)

In [7]:
t

(1, 2, 3)

In [8]:
t = 1, 2, 3

In [9]:
t

(1, 2, 3)

In [10]:
len(t)

3

In [11]:
t[0]

1

In [12]:
t[2]

3

In [13]:
numbers = {'one':1, 'two':2, 'three':3}

In [14]:
numbers['two']

2

In [15]:
numbers['ninety'] = 90

In [16]:
print(numbers)

{'one': 1, 'two': 2, 'three': 3, 'ninety': 90}


In [17]:
numbers['ninety']

90

In [18]:
primes = {2, 3, 5, 7}
odds = {1, 3, 5, 7, 9}

In [19]:
primes.union(odds)

{1, 2, 3, 5, 7, 9}

In [20]:
primes | odds

{1, 2, 3, 5, 7, 9}

In [21]:
primes & odds

{3, 5, 7}

In [22]:
primes.intersection(odds)

{3, 5, 7}

In [23]:
primes - odds

{2}

In [24]:
primes.difference(odds)

{2}

In [25]:
primes ^ odds

{1, 2, 9}

In [26]:
primes.symmetric_difference(odds)

{1, 2, 9}

# Control Flow

In [27]:
x = -15

if x == 0:
    print(x, "is zero")
elif x > 0:
    print(x, "is positive")
elif x < 0:
    print(x, "is negative")
else:
    print(x, "is unlike anything I've ever seen")
    

-15 is negative


In [28]:
for N in [2, 3, 5, 7]:
    print(N, end=' ')

2 3 5 7 

In [29]:
for i in range(10):
    print(i, end=' ')

0 1 2 3 4 5 6 7 8 9 

In [30]:
list(range(5, 10))

[5, 6, 7, 8, 9]

In [31]:
list(range(0, 10, 2))

[0, 2, 4, 6, 8]

In [3]:
i = 0
while i < 200:
    print(i, end=' ')
    i += 1

0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 64 65 66 67 68 69 70 71 72 73 74 75 76 77 78 79 80 81 82 83 84 85 86 87 88 89 90 91 92 93 94 95 96 97 98 99 100 101 102 103 104 105 106 107 108 109 110 111 112 113 114 115 116 117 118 119 120 121 122 123 124 125 126 127 128 129 130 131 132 133 134 135 136 137 138 139 140 141 142 143 144 145 146 147 148 149 150 151 152 153 154 155 156 157 158 159 160 161 162 163 164 165 166 167 168 169 170 171 172 173 174 175 176 177 178 179 180 181 182 183 184 185 186 187 188 189 190 191 192 193 194 195 196 197 198 199 

In [4]:
for n in range(20):
    # if the remainder of n / 2 is 0, skip the rest of the loop
    if n % 2 == 0:
        continue
    print(n, end=' ')

1 3 5 7 9 11 13 15 17 19 

In [7]:
#list all fibonacci numbers up to certain value

a, b = 0, 1
amax = 100
L = []

while True:
    (a, b) = (b, a + b)
    if a > amax:
        break
    L.append(a)

print(L)

[1, 1, 2, 3, 5, 8, 13, 21, 34, 55, 89]


In [8]:
#finding prime numbers within range

L = []
nmax = 30

for n in range(2, nmax):
    for factor in L:
        if n % factor == 0:
            break
    else: # no break
        L.append(n)
print(L)

[2, 3, 5, 7, 11, 13, 17, 19, 23, 29]


# Defining and Using Functions

In [9]:
print('abc')

abc


In [10]:
print(1, 2, 3)

1 2 3


In [12]:
print(1, 2, 3, sep='-')

1-2-3


In [16]:
def fibonacci(N, a=0, b=1):
    L = []
    while len(L) < N:
        a, b = b, a + b
        L.append(a)
    return L

In [17]:
fibonacci(10)

[1, 1, 2, 3, 5, 8, 13, 21, 34, 55]

In [18]:
fibonacci(10, 0, 2)

[2, 2, 4, 6, 10, 16, 26, 42, 68, 110]

In [19]:
fibonacci(10, b=3, a=1)

[3, 4, 7, 11, 18, 29, 47, 76, 123, 199]

In [20]:
def catch_all(*args, **kwargs):
    print("args =", args)
    print("kwargs =", kwargs)

In [21]:
catch_all(1, 2, 3, a=4, b=5)

args = (1, 2, 3)
kwargs = {'a': 4, 'b': 5}


In [22]:
catch_all('a', keyword=2)

args = ('a',)
kwargs = {'keyword': 2}


In [23]:
inputs = (1, 2, 3)
keywords = {'pi': 3.14}

catch_all(*inputs, **keywords)

args = (1, 2, 3)
kwargs = {'pi': 3.14}


In [24]:
add = lambda x, y: x + y
add(1, 2)

3

In [26]:
def add(x, y):
    return x + y
add(1, 2)

3

In [28]:
sorted([2,4,3,5,1,6])

[1, 2, 3, 4, 5, 6]

In [27]:
data = [{'first':'Guido', 'last':'Van Rossum', 'YOB':1956},
        {'first':'Grace', 'last':'Hopper',     'YOB':1906},
        {'first':'Alan',  'last':'Turing',     'YOB':1912}]

In [29]:
# sort alphabetically by first name
sorted(data, key=lambda item: item['first'])

[{'first': 'Alan', 'last': 'Turing', 'YOB': 1912},
 {'first': 'Grace', 'last': 'Hopper', 'YOB': 1906},
 {'first': 'Guido', 'last': 'Van Rossum', 'YOB': 1956}]

In [31]:
# sort by year of birth
sorted(data, key=lambda item: item['YOB'])

[{'first': 'Grace', 'last': 'Hopper', 'YOB': 1906},
 {'first': 'Alan', 'last': 'Turing', 'YOB': 1912},
 {'first': 'Guido', 'last': 'Van Rossum', 'YOB': 1956}]

# Errors and Exceptions

In [120]:
print(Q)

NameError: name 'Q' is not defined

In [121]:
1 + 'abc'

TypeError: unsupported operand type(s) for +: 'int' and 'str'

In [122]:
2 / 0

ZeroDivisionError: division by zero

In [123]:
L = [1, 2, 3]
L[1000]

IndexError: list index out of range

In [124]:
try:
    print("this gets executed first")
except:
    print("this gets executed only if there is an error")

this gets executed first


In [125]:
try:
    print("let's try something:")
    x = 1 / 0 # ZeroDivisionError
except:
    print("something bad happened!")

let's try something:
something bad happened!


In [129]:
def safe_divide(a, b):
    try:
        return a / b
    except:
        return 1E100

In [130]:
safe_divide(1, 2)

0.5

In [131]:
safe_divide(2, 0)

1e+100

In [132]:
safe_divide (1, '2')

1e+100

In [133]:
def safe_divide(a, b):
    try:
        return a / b
    except ZeroDivisionError:
        return 1E100

In [134]:
safe_divide(1, 0)

1e+100

In [135]:
safe_divide(1, '2')

TypeError: unsupported operand type(s) for /: 'int' and 'str'

In [136]:
raise RuntimeError("my error message")

RuntimeError: my error message

In [137]:
def fibonacci(N):
    L = []
    a, b = 0, 1
    while len(L) < N:
        a, b = b, a + b
        L.append(a)
    return L

In [138]:
def fibonacci(N):
    if N < 0:
        raise ValueError("N must be non-negative")
    L = []
    a, b = 0, 1
    while len(L) < N:
        a, b = b, a + b
        L.append(a)
    return L

In [139]:
fibonacci(10)

[1, 1, 2, 3, 5, 8, 13, 21, 34, 55]

In [140]:
fibonacci(-10)

ValueError: N must be non-negative

In [141]:
N = -10
try:
    print("trying this...")
    print(fibonacci(N))
except ValueError:
    print("Bad value: need to do something else")

trying this...
Bad value: need to do something else


In [142]:
try:
    x = 1 / 0
except ZeroDivisionError as err:
    print("Error class is:  ", type(err))
    print("Error message is:", err)

Error class is:   <class 'ZeroDivisionError'>
Error message is: division by zero


In [143]:
class MySpecialError(ValueError):
    pass

raise MySpecialError("here's the message")

MySpecialError: here's the message

In [144]:
try:
    print("do something")
    raise MySpecialError("[informative error message here]")
except MySpecialError:
    print("do something else")

do something
do something else


In [146]:
try:
    print("try something here")
except:
    print("this happens only if it fails")
else:
    print("this happens only if it succeeds")
finally:
    print("this happens no matter what")

try something here
this happens only if it succeeds
this happens no matter what


# Iterators

In [147]:
for i in range(10):
    print(i, end=' ')

0 1 2 3 4 5 6 7 8 9 

In [151]:
for value in [2, 4, 6, 8, 10]:
    # do some operation
    print(value + 1, end=' ')

3 5 7 9 11 

In [152]:
iter([2, 4, 6, 8, 10])

In [154]:
I = iter([2, 4, 6, 8, 10])

In [155]:
print(next(I))

2


In [156]:
print(next(I))

4


In [157]:
print(next(I))

6


In [158]:
range(10)

range(0, 10)

In [159]:
iter(range(10))

In [160]:
for i in range(10):
    print(i, end=" ")

0 1 2 3 4 5 6 7 8 9 

In [161]:
N = 10 ** 12
for i in range(N):
    if i >= 10: break
    print(i, end=', ')

0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 

In [162]:
from itertools import count

for i in count():
    if i >= 10:
        break
    print(i, end=', ')

0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 

In [163]:
# Often you need to iterate not only the values in an array, 
# but also keep track of the index. You might be tempted to do things this way:

L = [2, 4, 6, 8, 10]
for i in range(len(L)):
    print(i, L[i])

0 2
1 4
2 6
3 8
4 10


In [164]:
# but this is a lot easier

for i, val in enumerate(L):
    print(i, val)

0 2
1 4
2 6
3 8
4 10


In [167]:
#Zips iterators together

L = [2, 4, 6, 8, 10]
R = [3, 6, 9, 12, 15]
for lval, rval in zip(L, R):
    print(lval, rval)

2 3
4 6
6 9
8 12
10 15


In [165]:
# find the first 10 square numbers
square = lambda x: x ** 2
for val in map(square, range(10)):
    print(val, end=' ')

0 1 4 9 16 25 36 49 64 81 

# find values up to 10 for which x % 2 is zero
is_even = lambda x: x % 2 == 0
for val in filter(is_even, range(10)):
    print(val, end=' ')

In [169]:
#flexible arguments (*, **)

print(*range(10))

0 1 2 3 4 5 6 7 8 9


In [170]:
#trickier way to do the map example from above

print(*map(lambda x: x ** 2, range(10)))

0 1 4 9 16 25 36 49 64 81


In [171]:
L1 = (1, 2, 3, 4)
L2 = ('a', 'b', 'c', 'd')

In [174]:
z = zip(L1, L2)
print(*z)

(1, 'a') (2, 'b') (3, 'c') (4, 'd')


In [175]:
z = zip(L1, L2)
new_L1, new_L2 = zip(*z)
print(new_L1, new_L2)

(1, 2, 3, 4) ('a', 'b', 'c', 'd')


In [176]:
# permutations

from itertools import permutations
p = permutations(range(3))
print(*p)

(0, 1, 2) (0, 2, 1) (1, 0, 2) (1, 2, 0) (2, 0, 1) (2, 1, 0)


In [177]:
# combinations

from itertools import combinations
c = combinations(range(4), 2)
print(*c)

(0, 1) (0, 2) (0, 3) (1, 2) (1, 3) (2, 3)


In [178]:
# product

from itertools import product
p = product('ab', range(3))
print(*p)

('a', 0) ('a', 1) ('a', 2) ('b', 0) ('b', 1) ('b', 2)


# List Comprehensions

In [180]:
# numbers between 1 and 20 that cannot be divided by 3

[i for i in range(20) if i % 3 > 0]

[1, 2, 4, 5, 7, 8, 10, 11, 13, 14, 16, 17, 19]

In [181]:
# instead of all this crap

L = []
for n in range(12):
    L.append(n ** 2)
L

[0, 1, 4, 9, 16, 25, 36, 49, 64, 81, 100, 121]

In [182]:
# just use this

[n ** 2 for n in range(12)]

[0, 1, 4, 9, 16, 25, 36, 49, 64, 81, 100, 121]

In [183]:
# multiple iteration

[(i, j) for i in range(2) for j in range(3)]

[(0, 0), (0, 1), (0, 2), (1, 0), (1, 1), (1, 2)]

In [184]:
# you can add a conditional to the end of the expression

[val for val in range(20) if val % 3 > 0]

[1, 2, 4, 5, 7, 8, 10, 11, 13, 14, 16, 17, 19]

The expression (i % 3 > 0) evaluates to True unless val is divisible by 3. Again, the English language meaning can be immediately read off: "Construct a list of values for each value up to 20, but only if the value is not divisible by 3". Once you are comfortable with it, this is much easier to write – and to understand at a glance – than the equivalent loop syntax:

In [185]:
L = []
for val in range(20):
    if val % 3:
        L.append(val)
L

[1, 2, 4, 5, 7, 8, 10, 11, 13, 14, 16, 17, 19]

In [186]:
# conditionals on a value

val = -10
val if val >= 0 else -val

10

In [188]:
# in range of 20, negate values that are divisible by 2

[val if val % 2 else -val
 for val in range(20) if val % 3]

[1, -2, -4, 5, 7, -8, -10, 11, 13, -14, -16, 17, 19]

In [189]:
# set comprehension (can have no duplicates)

{n**2 for n in range(12)}

{0, 1, 4, 9, 16, 25, 36, 49, 64, 81, 100, 121}

In [190]:
{a % 3 for a in range(1000)}

{0, 1, 2}

In [192]:
# : creates a direct comprehension

{n:n**2 for n in range(6)}

{0: 0, 1: 1, 2: 4, 3: 9, 4: 16, 5: 25}

In [193]:
# parenthesis instead of square brackets = generator comprehension

(n**2 for n in range(12))

<generator object <genexpr> at 0x0000022C2E110AC8>

# Regular Expressions

In [26]:
s = "sdf\nsdf"

In [27]:
print(s)

sdf
sdf


In [29]:
s = "s\tdf\nsdf"

In [30]:
print(s)

s	df
sdf


In [31]:
s = r"s\tdf\nsdf"

In [32]:
print(s)

s\tdf\nsdf


In [33]:
path = "myself\newdir"

In [34]:
print(path)

myself
ewdir


In [35]:
path = r"myself\newdir"

In [36]:
print(path)

myself\newdir


In [37]:
import re

In [38]:
s = "Mr X's phone number is 123-444-3445 and Mrs. Y's is 432-566-4444"

# \d = any digits
# \w = any alphanumeric
# \s = any alpha character

In [39]:
exp = re.compile(r'\d\d\d-\d\d\d-\d\d\d\d')
exp.findall(s)

['123-444-3445', '432-566-4444']

In [40]:
exp = re.compile(r'\d{3}-\d{3}-\d{4}')
exp.findall(s)

['123-444-3445', '432-566-4444']

In [41]:
s = "Mr X's email is mrx@client.org and Mrs. Y is mrsy@client.ca"

In [42]:
exp = re.compile(r'\w+@\w+.\w+')
exp.findall(s)

['mrx@client.org', 'mrsy@client.ca']

In [43]:
exp = re.compile(r'\w+@\w+.\w{2,3}')
exp.findall(s)

['mrx@client.org', 'mrsy@client.ca']

In [44]:
line = 'the quick brown fox jumped over a lazy dog'
line.split()

['the', 'quick', 'brown', 'fox', 'jumped', 'over', 'a', 'lazy', 'dog']

In [45]:
import re
regex = re.compile('\s+')
regex.split(line)

['the', 'quick', 'brown', 'fox', 'jumped', 'over', 'a', 'lazy', 'dog']

In [46]:
regex = re.compile(r'\w\s\w')
regex.findall('the fox is 9 years old')

['e f', 'x i', 's 9', 's o']

In [47]:
exp.sub("***********",s)

"Mr X's email is *********** and Mrs. Y is ***********"

In [48]:
exp.split(s)

["Mr X's email is ", ' and Mrs. Y is ', '']

In [49]:
exp.search(s)

<re.Match object; span=(16, 30), match='mrx@client.org'>

In [50]:
exp.match(s)

The following table lists a few of these characters that are commonly useful:

| Character | Description                 |.| Character | Description                     |
|-----------|-----------------------------|-|-----------|---------------------------------|
| ``"\d"``  | Match any digit             |-| ``"\D"``  | Match any non-digit             |
| ``"\s"``  | Match any whitespace        |-| ``"\S"``  | Match any non-whitespace        |
| ``"\w"``  | Match any alphanumeric char |-| ``"\W"``  | Match any non-alphanumeric char |

This is *not* a comprehensive list or description; for more details, see Python's [regular expression syntax documentation](https://docs.python.org/3/library/re.html#re-syntax).

# Splitting around vowels

In [51]:
regex = re.compile('[aeiou]')
regex.split('consequential')

['c', 'ns', 'q', '', 'nt', '', 'l']

# Matching a number/letter combination

In [52]:
regex = re.compile('[A-Z][0-9]')
regex.findall('1043879, G2, H6')

['G2', 'H6']

The following is a table of the repetition markers available for use in regular expressions:

| Character | Description | Example |
|-----------|-------------|---------|
| ``?`` | Match zero or one repetitions of preceding  | ``"ab?"`` matches ``"a"`` or ``"ab"`` |
| ``*`` | Match zero or more repetitions of preceding | ``"ab*"`` matches ``"a"``, ``"ab"``, ``"abb"``, ``"abbb"``... |
| ``+`` | Match one or more repetitions of preceding  | ``"ab+"`` matches ``"ab"``, ``"abb"``, ``"abbb"``... but not ``"a"`` |
| ``{n}`` | Match ``n`` repetitions of preeeding | ``"ab{2}"`` matches ``"abb"`` |
| ``{m,n}`` | Match between ``m`` and ``n`` repetitions of preceding | ``"ab{2,3}"`` matches ``"abb"`` or ``"abbb"`` |

In [53]:
s = "Mr X's phone number is +123-444-3445 and Mrs. Y's is 432-566-4444"

In [54]:
exp = re.compile(r'\+?\d{3}-\d{3}-\d{4}')
exp.findall(s)

['+123-444-3445', '432-566-4444']

In [55]:
s = "Mr X's phone number is 123-444-3445 and Mrs. Y's is 432-566-4444"

# Grouping

In [56]:
exp = re.compile(r'(\d{3})-\d{3}-\d{4}')
exp.findall(s)

['123', '432']

In [57]:
exp = re.compile(r'(\d{3})-(\d{3}-\d{4})')
exp.findall(s)

[('123', '444-3445'), ('432', '566-4444')]

# Warm Up Solutions

1- How many seconds are in an hour?

In [1]:
60 * 60

3600

2- Assign the result from the previous task to a variable called seconds_per_hour

In [2]:
seconds_per_hour = 60 * 60

3- How many seconds are in a day? Use your seconds_per_hour variable

In [3]:
seconds_per_hour * 24

86400

4- Calculate seconds per day again, but this time save the result in a variable called seconds_per_day

In [6]:
seconds_per_day = seconds_per_hour * 24
seconds_per_day

86400

5- Divide seconds_per_day by seconds_per_hour. Use floating-point (/) division.

In [7]:
seconds_per_day / seconds_per_hour

24.0

6- Divide seconds_per_day by seconds_per_hour using integer (//) division. Did this number agree with the floating-point value from the previous question, aside from the final .0?

In [8]:
seconds_per_day // seconds_per_hour

24

Add the two numbers num1 and num2

In [9]:
num1 = 5
num2 = 7
num1 + num2

12

Add the two lists hard_skills and soft_skills

In [11]:
hard_skills = ["Java", "Python"]
soft_skills = ["Presentation", "Leadership"]
hard_skills + soft_skills

['Java', 'Python', 'Presentation', 'Leadership']

Can you add num1 to hard_skills?

In [12]:
num1 + hard_skills

TypeError: unsupported operand type(s) for +: 'int' and 'list'

How do we add a number to a list?

In [13]:
hard_skills + [num1]

['Java', 'Python', 5]

In [14]:
hard_skills.append(num1)
hard_skills

['Java', 'Python', 5]

Can you multiply num1 by hard_skills?

In [15]:
num1 * hard_skills

['Java',
 'Python',
 5,
 'Java',
 'Python',
 5,
 'Java',
 'Python',
 5,
 'Java',
 'Python',
 5,
 'Java',
 'Python',
 5]

Add the two words word1 and word2

In [16]:
word1 = "Java"
word2 = "Python"
word1 + word2

'JavaPython'

Can you add/concatenate num1 to word1?

In [17]:
word1 + str(num1)

'Java5'

Multiply num1 by word1. What do you get?

In [18]:
num1 * word1

'JavaJavaJavaJavaJava'

Compare num1 to num2, using == and >=

In [19]:
num1 == num2

False

In [20]:
num1 >= num2

False

Compare word1 to word2, using == and >=

In [21]:
word1 == word2

False

In [22]:
word1 >= word2

False

Compare hard_skills to soft_skills, using == and >=

In [23]:
hard_skills == soft_skills

False

In [24]:
hard_skills >= soft_skills

False

Concatenate the variables containing student name and grade.

The output should be formatted to be The grade of Sammy is 88.444%

In [25]:
student = "Sammy"
grade = 88.444

Below are three ways of concatenating the variables:

Using + to concatenate strings
Using formatted strings f"". This only works in Python 3.6 and higher
Using the function format

In [26]:
"The grade of " + student + " is " + str(grade) + "%"

'The grade of Sammy is 88.444%'

In [27]:
f"The grade of {student} is {grade}%"

'The grade of Sammy is 88.444%'

In [28]:
"The grade of {0} is {1}%".format(student, grade)

'The grade of Sammy is 88.444%'

Convert grade to int, re-format the variables.

The output should be formatted to be The grade of Sammy is 88%

There are two ways to do it:

convert grade to int
use format specification in the formatted string

In [29]:
"The grade of " + student + " is " + str(int(grade)) + "%"

'The grade of Sammy is 88%'

In [30]:
f"The grade of {student} is {grade:.0f}%"

'The grade of Sammy is 88%'

What is the type of student?

In [31]:
type(student)

str

What is the first character in student?

In [32]:
student[0]

'S'

What is the type of the first character in student?

In [33]:
type(student[0])

str

In [34]:
msg = "data analytics is the future"

Using string functions, convert `msg` to:
- upper case
- title case
- capitalize

In [35]:
msg2 = msg.upper()
msg2

'DATA ANALYTICS IS THE FUTURE'

In [36]:
msg.title()

'Data Analytics Is The Future'

In [37]:
msg.capitalize()

'Data analytics is the future'

In [38]:
quote1="""An optimist believes we live in the best world possible; 
a pessimist fears this is true."""

In [39]:
print(quote1)

An optimist believes we live in the best world possible; 
a pessimist fears this is true.


How many words does `quote1` contain?

In [40]:
len(quote1) #incorrect it counts the number of characters in the string

89

In [41]:
len(quote1.split()) #split to a list and count items

16

In [42]:
quote1.count(" ") + 1 #count whitespaces and add one

16

In [43]:
quote2 = """If you think nobody cares if you're alive, 
try missing a couple of car payments.
		-- Earl Wilson"""

In [44]:
print(quote2)

If you think nobody cares if you're alive, 
try missing a couple of car payments.
		-- Earl Wilson


How many lines are there in `quote2`?

In [45]:
len(quote2.splitlines())

3

In [46]:
len(quote2.split('\n'))

3

In [47]:
quote2.count('\n')+1

3

Extract the name of the quoted person in `quote2`

In [48]:
print(quote2)

If you think nobody cares if you're alive, 
try missing a couple of car payments.
		-- Earl Wilson


In [49]:
s1 = quote2.split('--')
s1

["If you think nobody cares if you're alive, \ntry missing a couple of car payments.\n\t\t",
 ' Earl Wilson']

In [50]:
s1[1].strip()

'Earl Wilson'

In [51]:
b=quote2.find('--')

In [52]:
quote2[b+3:len(quote2)] #b+3 is accounting for the two dashes and the whitespace

'Earl Wilson'

1- Assign the value 7 to the variable guess_me. Then, write the conditional tests (if, else, and elif) to print the string 'too low' if guess_me is less than 7, 'too high' if greater than 7, and 'just right' if equal to 7.

In [53]:
guess_me = 7

if guess_me == 7:
    print("just right")
elif guess_me < 7:
    print("too low")
else:
    print("too high")

just right


2- Assign the value 7 to the variable guess_me and the value 1 to the variable start. Write a while loop that compares start with guess_me. Print too low if start is less than guess me. If start equals guess_me, print 'found it!' and exit the loop. If start is greater than guess_me, print 'oops' and exit the loop. Increment start at the end of the loop.

In [55]:
guess_me = 7
start = 19

while True:
    if start < guess_me:
        print("too low")
    elif start == guess_me:
        print("found it")
        break
    else:
        print('oops')
        break
    start += 1

oops


3- Use a for loop to print the values of the list [3, 2, 1, 0].

In [56]:
for v in [3,2,1,0]:
    print(v, end = ' ')

3 2 1 0 

4- Extract email address from `msg` using a loop.
Think about breaking the text to words, then print
words that contain `@`

In [57]:
msg = """My name is James Bond and my email address
is james.bond@mi6.co.uk
Do not hesitate to write me. I'm a man for all seasons.
You can write to my secretary Miss.Moneypenny@mi6.co.uk
"""

In [58]:
words = msg.split()
emails = []
for word in words:
    if '@' in word: #word.find('@') != -1:
        emails += [word]
        #emails.append(word)
        #print(word)
emails

['james.bond@mi6.co.uk', 'Miss.Moneypenny@mi6.co.uk']

Define a function `mask` that takes a string and returns `*` repeated as many times as the length of the string

For example, `mask("342-33-4432")` returns `***********`

In [59]:
def mask(text):
    return '*' * len(text)

In [60]:
mask("342-33-4432")

'***********'

Define `mask` using a lambda expression

In [61]:
mask = lambda x: '*' * len(x)

In [62]:
mask("342-33-4432")

'***********'

Modify `mask` to add an optional parameter which is the `masking_character`, by default this parameter is `*`

*Note* you will not be able to use a lambda expression with a default parameter, you have to use `def`

In [63]:
def mask(text, masking_character='*'):
    return masking_character * len(text)

In [64]:
mask("342-33-4432")

'***********'

In [65]:
mask("342-33-4432", '#')

'###########'

Define a function `is_number` that takes a string and returns `True` if the string is one word that contains a number defined as having all characters as digits or `-`

In [66]:
def is_number(word):
    for letter in word: #you can iterate over string!
        if letter not in ["0", "1", "2", "3", "4", "5", "6", "7", "8", "9", "-"]:
            return False
    else: #means all letters are digits or -
        return True

In [67]:
#another way
def is_number(word):
    return word.replace("-", "").isnumeric()

In [68]:
is_number("333-333")

True

1- Create a list called years_list, starting with the year of your birth, and each year thereafter until the year of your fifth birthday. 
For example, if you were born in 1980. the list would be years_list = [1980, 1981, 1982, 1983, 1984, 1985]

In [69]:
years_list = [1980, 1981, 1982, 1983, 1984, 1985]

2- In which year in years_list was your third birthday? Remember, you were 0 years of age for your first year.

In [70]:
years_list[3]

1983

3- In which year in years_list were you the oldest?

In [72]:
years_list[-1]

1985

4- Make a list called things with these three strings as elements: "mozzarella", "cinderella", "salmonella".

In [82]:
things = ["mozzarella", "cinderella", "salmonella"]

5- Capitalize the element in things that refers to a person and then print the list. Did it change the element in the list?

In [83]:
things[1] = things[1].capitalize()
things

['mozzarella', 'Cinderella', 'salmonella']

6- Make the cheesy element of things all uppercase and then print the list.

In [84]:
things[0] = things[0].upper()
things

['MOZZARELLA', 'Cinderella', 'salmonella']

7- Delete the disease element from things, collect your Nobel Prize, and print the list.

In [ ]:
# The pop() method removes the item at the given index from the list and returns the removed item. ... 
# If the index passed to the method is not in range, it throws IndexError: pop index out of range exception.
# The pop() method returns the item present at the given index.

things.pop()
things

In [86]:
things = ["mozzarella", "cinderella", "salmonella"]
things = things[0:2]
things

['mozzarella', 'cinderella']

In [87]:
things = ["mozzarella", "cinderella", "salmonella"]
things[-1:]=[]
things

['mozzarella', 'cinderella']

8- Create a list called surprise with the elements "Groucho", "Chico", and "Harpo".

In [88]:
surprise = ["Groucho", "Chico",  "Harpo"]

9- Lowercase the last element of the surprise list, reverse it, and then capitalize it.

In [90]:
surprise[-1].lower()[::-1].capitalize()

'Oprah'

In [91]:
# or spell it in details
last_element = surprise[-1]
last_element_in_lower_case = last_element.lower()
reversed_element = last_element_in_lower_case[::-1]
reversed_element.capitalize()

'Oprah'

10- Make an English-to-French dictionary called e2f and print it. Here are your starter words: dog is chien, cat is chat, and walrus is morse.

In [93]:
e2f = {'dog':'chien', 'cat':'chat', 'walrus':'morse'}

11- Using your three-word dictionary e2f, print the French word for walrus.

In [95]:
print(e2f['walrus'])

morse


12- Make a French-to-English dictionary called f2e from e2f. Use the items method.

In [96]:
e2f.items()

dict_items([('dog', 'chien'), ('cat', 'chat'), ('walrus', 'morse')])

In [97]:
# each item is a tuple
f2e = {}
for item in e2f.items():
    key = item[0]
    value = item[1]
    f2e[value]=key
f2e

{'chien': 'dog', 'chat': 'cat', 'morse': 'walrus'}

In [98]:
#another way to do it
f2e = {}
for k in e2f:
    v = e2f[k]
    f2e[v]=k
f2e

{'chien': 'dog', 'chat': 'cat', 'morse': 'walrus'}

13- Using f2e, print the English equivalent of the French word chien.

In [99]:
f2e['chien']

'dog'

14- Make and print a set of English words from the keys in e2f.

In [100]:
english_words = set(e2f.keys())
english_words

{'cat', 'dog', 'walrus'}

15- Make a multilevel dictionary called life. Use these strings for the topmost keys: 'animals', 'plants', and 'other'. Make the 'animals' key refer to another dictionary with the keys 'cats', 'octopi', and 'emus'. Make the 'cats' key refer to a list of strings with the values 'Henri', 'Grumpy', and 'Lucy'. Make all the other keys refer to empty dictionaries.

In [101]:
life = {'animals': 
            {
                'cats': ['Henri', 'Grumpy', 'Lucy'],
                'octopi': {},
                'emus': {}
            }, 
        'plants': {}, 
        'other': {}
       }

In [102]:
life

{'animals': {'cats': ['Henri', 'Grumpy', 'Lucy'], 'octopi': {}, 'emus': {}},
 'plants': {},
 'other': {}}

16- Print the top-level keys of life.

In [103]:
life.keys()

dict_keys(['animals', 'plants', 'other'])

In [104]:
for k in life.keys():
    print(k)

animals
plants
other


17- Print the keys for life['animals'].

In [105]:
life['animals'].keys()

dict_keys(['cats', 'octopi', 'emus'])

In [106]:
for k in life['animals'].keys():
    print(k)

cats
octopi
emus


18- Print the values for life['animals']['cats'].

In [108]:
life['animals']['cats']

['Henri', 'Grumpy', 'Lucy']

1- Write a list comprehensions that contains numbers from 1 to 31, append the prefix "1-" to the numbers, these are day of January

In [111]:
L = [f"1-{day}" for day in range(1,32)]

2- convert the list to a string so each entry prints on one line. Print the result

In [112]:
line = '\n'.join(L)
print(line)

1-1
1-2
1-3
1-4
1-5
1-6
1-7
1-8
1-9
1-10
1-11
1-12
1-13
1-14
1-15
1-16
1-17
1-18
1-19
1-20
1-21
1-22
1-23
1-24
1-25
1-26
1-27
1-28
1-29
1-30
1-31


In [113]:
from functools import reduce
line = reduce(lambda x, y: f"{x}\n{y}", L)
print(line)

1-1
1-2
1-3
1-4
1-5
1-6
1-7
1-8
1-9
1-10
1-11
1-12
1-13
1-14
1-15
1-16
1-17
1-18
1-19
1-20
1-21
1-22
1-23
1-24
1-25
1-26
1-27
1-28
1-29
1-30
1-31


3- Update the comprehension to vary prefix from 1- to 12- to generate all days of the year, do not worry about incorrect dates for now

In [114]:
L = [ f"{month}-{day}" for month in range(1, 13) for day in range(1,32) ]
L

['1-1',
 '1-2',
 '1-3',
 '1-4',
 '1-5',
 '1-6',
 '1-7',
 '1-8',
 '1-9',
 '1-10',
 '1-11',
 '1-12',
 '1-13',
 '1-14',
 '1-15',
 '1-16',
 '1-17',
 '1-18',
 '1-19',
 '1-20',
 '1-21',
 '1-22',
 '1-23',
 '1-24',
 '1-25',
 '1-26',
 '1-27',
 '1-28',
 '1-29',
 '1-30',
 '1-31',
 '2-1',
 '2-2',
 '2-3',
 '2-4',
 '2-5',
 '2-6',
 '2-7',
 '2-8',
 '2-9',
 '2-10',
 '2-11',
 '2-12',
 '2-13',
 '2-14',
 '2-15',
 '2-16',
 '2-17',
 '2-18',
 '2-19',
 '2-20',
 '2-21',
 '2-22',
 '2-23',
 '2-24',
 '2-25',
 '2-26',
 '2-27',
 '2-28',
 '2-29',
 '2-30',
 '2-31',
 '3-1',
 '3-2',
 '3-3',
 '3-4',
 '3-5',
 '3-6',
 '3-7',
 '3-8',
 '3-9',
 '3-10',
 '3-11',
 '3-12',
 '3-13',
 '3-14',
 '3-15',
 '3-16',
 '3-17',
 '3-18',
 '3-19',
 '3-20',
 '3-21',
 '3-22',
 '3-23',
 '3-24',
 '3-25',
 '3-26',
 '3-27',
 '3-28',
 '3-29',
 '3-30',
 '3-31',
 '4-1',
 '4-2',
 '4-3',
 '4-4',
 '4-5',
 '4-6',
 '4-7',
 '4-8',
 '4-9',
 '4-10',
 '4-11',
 '4-12',
 '4-13',
 '4-14',
 '4-15',
 '4-16',
 '4-17',
 '4-18',
 '4-19',
 '4-20',
 '4-21',
 '4-22',
 

4- now, address the issue of some months being only 30 days and february if 28 days (this year). Hint, use the valid_date function

In [115]:
def valid_date(day, month):
    days_of_month = {1:31, 2:28, 3:31, 4:30, 5:31, 6:30, 
                7:31, 8:31, 9:30, 10:31, 11:30, 12:31}
    max_month = days_of_month[month]
    return day <= max_month

In [116]:
L = [ f"{month}-{day}" for month in range(1, 13) for day in range(1,32) if valid_date(day, month)]
L

['1-1',
 '1-2',
 '1-3',
 '1-4',
 '1-5',
 '1-6',
 '1-7',
 '1-8',
 '1-9',
 '1-10',
 '1-11',
 '1-12',
 '1-13',
 '1-14',
 '1-15',
 '1-16',
 '1-17',
 '1-18',
 '1-19',
 '1-20',
 '1-21',
 '1-22',
 '1-23',
 '1-24',
 '1-25',
 '1-26',
 '1-27',
 '1-28',
 '1-29',
 '1-30',
 '1-31',
 '2-1',
 '2-2',
 '2-3',
 '2-4',
 '2-5',
 '2-6',
 '2-7',
 '2-8',
 '2-9',
 '2-10',
 '2-11',
 '2-12',
 '2-13',
 '2-14',
 '2-15',
 '2-16',
 '2-17',
 '2-18',
 '2-19',
 '2-20',
 '2-21',
 '2-22',
 '2-23',
 '2-24',
 '2-25',
 '2-26',
 '2-27',
 '2-28',
 '3-1',
 '3-2',
 '3-3',
 '3-4',
 '3-5',
 '3-6',
 '3-7',
 '3-8',
 '3-9',
 '3-10',
 '3-11',
 '3-12',
 '3-13',
 '3-14',
 '3-15',
 '3-16',
 '3-17',
 '3-18',
 '3-19',
 '3-20',
 '3-21',
 '3-22',
 '3-23',
 '3-24',
 '3-25',
 '3-26',
 '3-27',
 '3-28',
 '3-29',
 '3-30',
 '3-31',
 '4-1',
 '4-2',
 '4-3',
 '4-4',
 '4-5',
 '4-6',
 '4-7',
 '4-8',
 '4-9',
 '4-10',
 '4-11',
 '4-12',
 '4-13',
 '4-14',
 '4-15',
 '4-16',
 '4-17',
 '4-18',
 '4-19',
 '4-20',
 '4-21',
 '4-22',
 '4-23',
 '4-24',
 '4-25',
 

5- using dictionary comprehensions create `f2e` dictionary from the `e2f` dictionary

In [117]:
e2f = {'hi':'bonjour', 'bye':'au revoir', 'bread':'pain', 'water':'eau'}

In [118]:
f2e = { e2f[k]:k for k in e2f}
f2e

{'bonjour': 'hi', 'au revoir': 'bye', 'pain': 'bread', 'eau': 'water'}

In [119]:
f2e = {item[1]:item[0] for item in e2f.items()}
f2e

{'bonjour': 'hi', 'au revoir': 'bye', 'pain': 'bread', 'eau': 'water'}

In [120]:
f2e = {v:k for (k,v) in e2f.items()}
f2e

{'bonjour': 'hi', 'au revoir': 'bye', 'pain': 'bread', 'eau': 'water'}

In [58]:
email = """Hi Jane, 
I hope you are doing well. As I explained in my previous message, 
Mr. James' social is 134-44-3212. His email is james.bond@secret.org
Please get in touch with him ASAP.

Also, Mr. Bond will be dribing his Aston Martin, it has a 3-way revolving front number 
plate showing LU-6789, 4711-EA-62 or BMT-216A. Please ensure paying tolls
for all. If you have any questions, please contact me: +44 (0)345 363 3630

"""

Using regular expressions, extract from the email
- numbers (e.g. social and phone numbers)
- email address
- license plates

In [59]:
import re
social = re.compile('\d+-\d+-\d+')
social.findall(email)

['134-44-3212']

In [60]:
phone = re.compile('\+\d+[ ()\d]+')
phone.findall(email)

['+44 (0)345 363 3630']

In [61]:
email_ = re.compile('\w+[.\w]+@\w+\.\w{3}')
email_.findall(email)

['james.bond@secret.org']

In [62]:
license = re.compile('[\d-]*[A-Z]+-[\dA-Z]+')
license.findall(email)

['LU-6789', '4711-EA-62', 'BMT-216A']

Blank out this sensitive information using *

In [63]:
for regex in [social, phone, email_, license]:
    email = regex.sub('******', email)
print(email)

Hi Jane, 
I hope you are doing well. As I explained in my previous message, 
Mr. James' social is ******. His email is ******
Please get in touch with him ASAP.

Also, Mr. Bond will be dribing his Aston Martin, it has a 3-way revolving front number 
plate showing ******, ****** or ******. Please ensure paying tolls
for all. If you have any questions, please contact me: ******




# Workout Solutions

Convert the [MIS faculty directory](https://www.terry.uga.edu/directory/mis/) to Markdown

*You don't need to convert the whole table just few entries*

|                                                          Image                                                          |         Name         |      Title       |    Type     |    Phone     |                    Email                    |                  Office                   |
| ----------------------------------------------------------------------------------------------------------------------- | -------------------- | ---------------- | ----------- | ------------ | ------------------------------------------- | ----------------------------------------- |
| ![](http://directory.domain-account.com/_resources/images/directory/faculty/Karen%20Aguar.jpg)                          | Karen Aguar          | Lecturer         | Faculty     | .            | [kaguar@uga.edu](mailto:kaguar@uga.edu)     | 630 South Lumpkin Street C428 Benson Hall |
| ![](http://directory.domain-account.com/_resources/images/directory/faculty/JanineAronson_29143-014_portrait_large.jpg) | Janine Elyse Aronson | Professor        | Faculty     | 706-542-0991 | [jaronson@uga.edu](mailto:jaronson@uga.edu) | 630 South Lumpkin Street C418 Benson Hall |
| ![](http://directory.domain-account.com/_resources/images/directory/students/eric%20bogert_portrait_large.jpg)          | Eric Bogert          | Doctoral Student | PhD Student | .            | [etbogert@uga.edu](mailto:etbogert@uga.edu) | 620 South Lumpkin Street B423 Amos Hall   |


1- Enter the following into Python, assigning each to a unique variable name:
- 4
- 3.1415
- 1.0
- 2+4j 
- 'Hello'
- 'World'

In [121]:
num  = 4
pi   = 3.1415
one  = 1.0
comp = 2 + 4j
str1 = 'Hello'
str2 = 'World'

2- What is the type of each variable? Use _type_ if you aren’t sure.

In [122]:
type(num)

int

In [123]:
type(pi)

float

In [124]:
type(one)

float

In [125]:
type(one)

float

In [126]:
type(comp)

complex

In [127]:
type(str1)

str

In [128]:
type(str2)

str

3- Try the following operators with the variables. Which of the 6 types can be:
- Added +
- Subtracted -
- Multiplies *
- Divided /

In [129]:
num  = 4
pi   = 3.1415
one  = 1.0
comp = 2 + 4j
str1 = 'Hello'
str2 = 'World'

It is possible to combine numerical types with mathematical operations

In [130]:
comp - num + pi * one / comp

(-1.6858499999999998+3.3716999999999997j)

It is possible to multiply a string by an integer to repeat it

In [131]:
num * str1

'HelloHelloHelloHello'

It is possible to add two strings to concatenate them

In [132]:
str1 + str2

'HelloWorld'

4- What are the types of the output (when an error is not produced) in the above operations?

In [133]:
type(comp - num + pi * one / comp)

complex

In [134]:
type(num * str1)

str

In [135]:
type(str1 + str2)

str

5- What are the numerator and denominator of the number 3.5

*Hint* use the function `as_integer_ratio()`

In [136]:
n = 3.5
n.as_integer_ratio()

(7, 2)

The code below generates a random number between 1 and 10. 
Write a boolean expression that evaluates to `True` when the number is odd

- Try different ways to write the expression, using boolean operations and membership operators

In [161]:
import random
ran = random.randint(1, 10)
ran

7

In [162]:
ran % 2 == 1

True

In [163]:
ran in [1, 3, 5, 7] 

True

![Play Tennis Decision Tree](https://nullpointerexception1.files.wordpress.com/2017/11/decision-tree-e1513448957591.jpg)

Write a boolean expression that represents the decision tree based on the variables below

In [164]:
outlook = input() #Sunny, Overcast, Rain
humidity = input() #High, Normal
wind = input() #Strong, Weak

Overcast
Normal
Weak


In [165]:
(outlook == "Sunny" and humidity == "Normal") or (outlook == "Overcast") or (outlook == "Rain" and wind == "Weak")

True

Concatenate the variables containing student name and grade.

Keep only one decimal digit.

The output should be formatted to be `The grade of Sammy is 88.4%`

In [166]:
student = "Sammy"
grade = 88.444

Solution 1: Use round function to round grade

In [167]:
rounded_grade = round(grade, 1)

In [168]:
f"The grade of {student} is {rounded_grade}%"

'The grade of Sammy is 88.4%'

Solution 2: Use formatted string and provide a format sepcification

In [170]:
f"The grade of {student} is {rounded_grade:.1f}%"

'The grade of Sammy is 88.4%'

In [171]:
msg = """My name is James Bond and my email address
is james.bond@mi6.co.uk
Do not hesitate to write me. I'm a man for all seasons
"""

In [172]:
print(msg)

My name is James Bond and my email address
is james.bond@mi6.co.uk
Do not hesitate to write me. I'm a man for all seasons



Extract the email address in the string `msg`

**Solution 1**

Use string functions

In [173]:
# convert new line to a space
msg2 = msg.replace('\n', ' ')
# first locate the location of the @
at_index = msg2.find('@')
# next find the white space after the @
space_right = msg2.find(' ', at_index)
# next find the white space before the @
space_left = msg2.rfind(' ', 0, at_index)

In [174]:
space_left, at_index, space_right


(45, 56, 66)

In [175]:
bond_email=msg2[space_left+1:space_right]
bond_email

'james.bond@mi6.co.uk'

**Solution 2**

Split to words and locate words with `@`

In [177]:
words = msg.split()
words

['My',
 'name',
 'is',
 'James',
 'Bond',
 'and',
 'my',
 'email',
 'address',
 'is',
 'james.bond@mi6.co.uk',
 'Do',
 'not',
 'hesitate',
 'to',
 'write',
 'me.',
 "I'm",
 'a',
 'man',
 'for',
 'all',
 'seasons']

In [178]:
bond_email = None
for word in words:
    if '@' in word:
        bond_email = word
bond_email

'james.bond@mi6.co.uk'

In [179]:
the_cow_said="""
____________________________________ 
/ "An ounce of prevention is worth a \
\ pound of purge."                   /
 ------------------------------------ 
        \   ^__^
         \  (oo)\_______
            (__)\       )\/\
                ||----w |
                ||     ||
"""

Extract the wisdom that the cow said

**Solution 1**

In [181]:
# The quote starts with " and ends with "

In [182]:
first_quote = the_cow_said.find('"')
second_quote = the_cow_said.rfind('"')

In [183]:
wisdom = the_cow_said[first_quote+1:second_quote]
wisdom

'An ounce of prevention is worth a \\ pound of purge.'

In [186]:
# clean the wisdom from \\
wisdom = wisdom.replace(' \\ ', '')
wisdom

'An ounce of prevention is worth a  pound of purge.'

**Solution 2**

In [187]:
sentences = the_cow_said.split('"')
sentences

['\n____________________________________ \n/ ',
 'An ounce of prevention is worth a \\ pound of purge.',
 '                   /\n ------------------------------------ \n        \\   ^__^\n         \\  (oo)\\_______\n            (__)\\       )\\/                ||----w |\n                ||     ||\n']

In [188]:
wisdom = sentences[1].replace('\\', '')
wisdom

'An ounce of prevention is worth a  pound of purge.'

0- use a loop to extract numbers from the secret message below.
Note that numbers can contain the extra character `-`

Write a loop to go over the words in the secrect message.
Then write an inner loop to go over each character in the word.
If all characters are numbers or `-` then the word is a number.
Print it!

In [189]:
msg = """some secret information
the client number is 444-222-5555 and her social
is 345-22-5556"""

In [190]:
i = 0
while i < 10:
    print(i)
    i += 1
else:
    print("i is ten")

0
1
2
3
4
5
6
7
8
9
i is ten


In [191]:
words = msg.split()
for word in words:
    #print(word)
    sensitive = True
    for c in word:
        if c not in '0123456789-':
            sensitive = False
            break
    if sensitive:
        print(word)

444-222-5555
345-22-5556


In [192]:
words = msg.split()
for word in words:
    for c in word:
        if c not in '0123456789-':
            break
    else:
        print(word)

444-222-5555
345-22-5556


In [193]:
words

['some',
 'secret',
 'information',
 'the',
 'client',
 'number',
 'is',
 '444-222-5555',
 'and',
 'her',
 'social',
 'is',
 '345-22-5556']

1- Write a loop that prints numbers from 1 to 31, append the prefix "1-" to the numbers, these are day of January, see sample output below

In [194]:
1-1
1-2
1-3
1-4
1-5
1-6
1-7
1-8
1-9
1-10
1-11
1-12
1-13
1-14
1-15
1-16
1-17
1-18
1-19
1-20
1-21
1-22
1-23
1-24
1-25
1-26
1-27
1-28
1-29
1-30
1-31

-30

In [195]:
for day in range(1, 32):
    print(f"1-{day}")

1-1
1-2
1-3
1-4
1-5
1-6
1-7
1-8
1-9
1-10
1-11
1-12
1-13
1-14
1-15
1-16
1-17
1-18
1-19
1-20
1-21
1-22
1-23
1-24
1-25
1-26
1-27
1-28
1-29
1-30
1-31


2- Vary the prefix from 1- to 12- to print all days of the year, do not worry about incorrect dates for now

In [196]:
for month in range(1,13):
    for day in range(1, 32):
        print(f"{month}-{day}")
    print(10*"-")

1-1
1-2
1-3
1-4
1-5
1-6
1-7
1-8
1-9
1-10
1-11
1-12
1-13
1-14
1-15
1-16
1-17
1-18
1-19
1-20
1-21
1-22
1-23
1-24
1-25
1-26
1-27
1-28
1-29
1-30
1-31
----------
2-1
2-2
2-3
2-4
2-5
2-6
2-7
2-8
2-9
2-10
2-11
2-12
2-13
2-14
2-15
2-16
2-17
2-18
2-19
2-20
2-21
2-22
2-23
2-24
2-25
2-26
2-27
2-28
2-29
2-30
2-31
----------
3-1
3-2
3-3
3-4
3-5
3-6
3-7
3-8
3-9
3-10
3-11
3-12
3-13
3-14
3-15
3-16
3-17
3-18
3-19
3-20
3-21
3-22
3-23
3-24
3-25
3-26
3-27
3-28
3-29
3-30
3-31
----------
4-1
4-2
4-3
4-4
4-5
4-6
4-7
4-8
4-9
4-10
4-11
4-12
4-13
4-14
4-15
4-16
4-17
4-18
4-19
4-20
4-21
4-22
4-23
4-24
4-25
4-26
4-27
4-28
4-29
4-30
4-31
----------
5-1
5-2
5-3
5-4
5-5
5-6
5-7
5-8
5-9
5-10
5-11
5-12
5-13
5-14
5-15
5-16
5-17
5-18
5-19
5-20
5-21
5-22
5-23
5-24
5-25
5-26
5-27
5-28
5-29
5-30
5-31
----------
6-1
6-2
6-3
6-4
6-5
6-6
6-7
6-8
6-9
6-10
6-11
6-12
6-13
6-14
6-15
6-16
6-17
6-18
6-19
6-20
6-21
6-22
6-23
6-24
6-25
6-26
6-27
6-28
6-29
6-30
6-31
----------
7-1
7-2
7-3
7-4
7-5
7-6
7-7
7-8
7-9
7-10
7-11
7-12
7-13
7-

3- now, address the issue of some months being only 30 days and february if 29 days (this year)

In [197]:
for month in range(1,13):
    
    if month in [4,6,9,11]:
        days = 30
    elif month == 2:
        days = 29
    else:
        days = 31
        
    for day in range(1, days + 1):
        print(f"{month}-{day}")
    print(10*"-")

1-1
1-2
1-3
1-4
1-5
1-6
1-7
1-8
1-9
1-10
1-11
1-12
1-13
1-14
1-15
1-16
1-17
1-18
1-19
1-20
1-21
1-22
1-23
1-24
1-25
1-26
1-27
1-28
1-29
1-30
1-31
----------
2-1
2-2
2-3
2-4
2-5
2-6
2-7
2-8
2-9
2-10
2-11
2-12
2-13
2-14
2-15
2-16
2-17
2-18
2-19
2-20
2-21
2-22
2-23
2-24
2-25
2-26
2-27
2-28
2-29
----------
3-1
3-2
3-3
3-4
3-5
3-6
3-7
3-8
3-9
3-10
3-11
3-12
3-13
3-14
3-15
3-16
3-17
3-18
3-19
3-20
3-21
3-22
3-23
3-24
3-25
3-26
3-27
3-28
3-29
3-30
3-31
----------
4-1
4-2
4-3
4-4
4-5
4-6
4-7
4-8
4-9
4-10
4-11
4-12
4-13
4-14
4-15
4-16
4-17
4-18
4-19
4-20
4-21
4-22
4-23
4-24
4-25
4-26
4-27
4-28
4-29
4-30
----------
5-1
5-2
5-3
5-4
5-5
5-6
5-7
5-8
5-9
5-10
5-11
5-12
5-13
5-14
5-15
5-16
5-17
5-18
5-19
5-20
5-21
5-22
5-23
5-24
5-25
5-26
5-27
5-28
5-29
5-30
5-31
----------
6-1
6-2
6-3
6-4
6-5
6-6
6-7
6-8
6-9
6-10
6-11
6-12
6-13
6-14
6-15
6-16
6-17
6-18
6-19
6-20
6-21
6-22
6-23
6-24
6-25
6-26
6-27
6-28
6-29
6-30
----------
7-1
7-2
7-3
7-4
7-5
7-6
7-7
7-8
7-9
7-10
7-11
7-12
7-13
7-14
7-15
7-16
7-17
7-

In [198]:
msg = """some secret information
the client number is 444-222-5555 and her social
is 345-22-5556 email is precious.client@wealth.org
other than that the weather is sunny and life is good"""

Write a function `classify` that takes a string and classifies sensitive information. Use the functions `is_number` and `mask` from the warmup exercise.

For example, classifying `msg` should returns

`some secret information
the client number is ************ and her social
is *********** email is precious.client@wealth.org
other than that the weather is sunny and life is good`

In [199]:
def mask(text, masking_character='*'):
    return masking_character * len(text)

In [200]:
def is_number(word):
    for letter in word: #you can iterate over string!
        if letter not in ["0", "1", "2", "3", "4", "5", "6", "7", "8", "9", "-"]:
            return False
    else: #means all letters are digits or -
        return True

In [201]:
def classify(msg):
    words = msg.split()
    for word in words:
        if is_number(word):
            msg = msg.replace(word, mask(word))
    return msg

In [202]:
print(classify(msg))

some secret information
the client number is ************ and her social
is *********** email is precious.client@wealth.org
other than that the weather is sunny and life is good


The last function will take care of sensitive numbers, but we would like to classify other sensitive information such as emails. I provide below a function `is_email` that returns `True` if the parameter is an email address

Modify, the function `classify` to take an optional parameter, which is the function to use to detect sensitive information (e.g., is_number, is_email)

Use the new function along with `is_number` and `is_email` to classify `msg`

For example, `print(classifiy(msg, is_number))` should print

`some secret information
the client number is ************ and her social
is *********** email is precious.client@wealth.org
other than that the weather is sunny and life is good`

and `print(classifiy(msg, is_email))` should print

`some secret information
the client number is 444-222-5555 and her social
is 345-22-5556 email is **************************
other than that the weather is sunny and life is good`

Can you classify both numbers and emails?

In [203]:
def is_email(text):
    return '@' in text and not text.startswith('@')

In [204]:
def classify(msg, is_sensitive = is_number):
    words = msg.split()
    for word in words:
        if is_sensitive(word):
            msg = msg.replace(word, mask(word))
    return msg

In [205]:
print(classify(msg))

some secret information
the client number is ************ and her social
is *********** email is precious.client@wealth.org
other than that the weather is sunny and life is good


In [206]:
print(classify(msg, is_email))

some secret information
the client number is 444-222-5555 and her social
is 345-22-5556 email is **************************
other than that the weather is sunny and life is good


In [207]:
print(classify(classify(msg, is_email)))

some secret information
the client number is ************ and her social
is *********** email is **************************
other than that the weather is sunny and life is good


In [208]:
print(classify(msg, lambda x: is_email(x) or is_number(x)))

some secret information
the client number is ************ and her social
is *********** email is **************************
other than that the weather is sunny and life is good


Many news analytics tools aggregate content from popular news outlet and highlight trending and important topics. As an example, check out [Infomous](https://www.infomous.com)

Let us build a mini news analytics tool. Using CNN top news stories on `9/3/2019`

- How many words does `article` contain?
- How many unique words?
- Sort these words by their frequency
- Some of the words have high frequency because they are common words in english, the list `stop_words` contains a list of these words. Exclude words in `stop_words` from your report of the above

In [209]:
article = """Hurricane Dorian leaves terrible damage and stranded residents in Bahamas
While residents evacuate, these men and women are flying straight into Hurricane Dorian
Here's what Hurricane Dorian is expected to do as it crawls toward the US
Bahamas' tourism could be devastated for a long time after Hurricane Dorian
Why Donald Trump golfing during Hurricane Dorian is a problem
Opinion: Donald Trump isn't up to the job
Quickly catch up on the day's news
Pentagon diverts $3.6 billion in military construction funds to build Trump's border wall
These were the passengers and crew members on board the California dive boat
What we know about the California dive boat and its last excursion
Search suspended in the deadly dive-boat fire off California coast
Opinion: The diving world tries to come to grips with devastating fire
Here's the real reason Mike Pence is staying at Donald Trump's hotel in Ireland
Want to live longer? You may want to ditch these drinks
Judge rules White House must give Playboy columnist Brian Karem his press pass back
Boris Johnson has just taken a huge gamble over Brexit
Brexit rebels seize Parliament's agenda in major blow to Boris Johnson
The UK could be heading for an early election. Here's what you need to know
What is a no-deal Brexit and what would it mean for Britain?
West Texas shooter bought gun in private sale
Teenage boy goes blind after existing on Pringles and french fries
A 14-year-old confessed to killing all five of his family members in an Alabama home, authorities say
Dorian, Comey and Debra Messing: What Trump tweeted on Labor Day weekend
Kroger asks customers not to openly carry guns in its stores
Walmart ends all handgun ammunition sales and asks customers not to carry guns into stores
Walmart CEO implores Congress to 'do their part' to stop gun violence
McConnell says he won't take up gun bill unless Trump says he will sign it
Police: Driver charged in abduction says child was sold for $10,000
Kristen Stewart says she was told not to hold hands with her girlfriend in public
Ariana Grande sues Forever 21 over ads featuring 'look-alike model'
Chicago mayor to Ted Cruz over city shootings: 'Keep our name out of your mouth'
An airline employee thought two men at Newark airport looked suspicious so she yelled for people to evacuate
Detroit Tigers prospect dies after electric skateboard accident
Mattis dodges when asked if he will speak candidly about Trump before 2020 election
Biden's campaign is already making Iowa excuses
Elizabeth Warren embraces Jay Inslee's climate change platform
Joe Manchin announces he'll stay in Senate and won't run for West Virginia governor
1 quote that perfectly explains the 2020 Democratic primary
This is why Congress remains deadlocked on climate and guns
Popeyes customer pulls a gun after being told there were no more chicken sandwiches
Robert Pattinson was blanking 'furious' over Batman leak
Justin Bieber shares use of 'heavy drugs' in revealing post
Cherokee Nation names first ever delegate to Congress
US, UK, France and Iran may be complicit in Yemen war crimes, UN says
Loud 'boom' reported across Central New York was probably a fireball entering Earth's atmosphere
Man arrested for buying ticket just to wave his wife off at gate
The problem of an 'infestation' of travel influencers
'It Chapter Two' doesn't know when to close the book
Leslie Jones bids 'Saturday Night Live' farewell
Days after her brother was charged in a triple slaying, Simone Biles says her heart aches 'especially for the victims'
Actor Cuba Gooding Jr.'s sex abuse trial has been delayed
Emma Thompson: Everything depends on what we do now"""

In [210]:
words = article.split()

In [211]:
#How many words does article contain?
len(words)

606

In [212]:
#How many unique words?
unique_words = set(words)
len(unique_words)

418

In [213]:
#Sort these words by their frequency
word_freq = {}
# first initialize the dictionary
for word in unique_words:
    word_freq[word]=0
# second, process text, increase count for each word
for word in words:
    word_freq[word]+=1

In [214]:
word_freq

{'fireball': 1,
 "Biden's": 1,
 'You': 1,
 "Parliament's": 1,
 'Comey': 1,
 'sale': 1,
 'Congress': 3,
 'White': 1,
 'says': 6,
 'House': 1,
 'Driver': 1,
 'Joe': 1,
 "Jr.'s": 1,
 "drugs'": 1,
 'Pentagon': 1,
 "Earth's": 1,
 'people': 1,
 "model'": 1,
 'embraces': 1,
 'sues': 1,
 'Iowa': 1,
 'Iran': 1,
 'Simone': 1,
 'being': 1,
 'dodges': 1,
 'dive-boat': 1,
 'Donald': 3,
 'golfing': 1,
 'agenda': 1,
 'could': 2,
 'post': 1,
 'reason': 1,
 'Quickly': 1,
 'give': 1,
 'Central': 1,
 'Karem': 1,
 'military': 1,
 'public': 1,
 'ends': 1,
 "Two'": 1,
 "he'll": 1,
 'construction': 1,
 'straight': 1,
 'Justin': 1,
 'Mike': 1,
 'our': 1,
 'election': 1,
 'Stewart': 1,
 'hotel': 1,
 'ever': 1,
 'electric': 1,
 'job': 1,
 'farewell': 1,
 'an': 3,
 'Biles': 1,
 'catch': 1,
 'members': 2,
 'violence': 1,
 "'Keep": 1,
 'now': 1,
 'confessed': 1,
 'about': 2,
 "Inslee's": 1,
 'shares': 1,
 '$10,000': 1,
 'longer?': 1,
 'want': 1,
 'entering': 1,
 'with': 2,
 'girlfriend': 1,
 'bids': 1,
 'Warren': 

In [215]:
#sort the dictionary

# first get a list of items
items = list(word_freq.items())

#items is a list, we can sort usint .sort or sorted
#however, we need to specify how the sort to be done

In [216]:
#take a look at the first item in items
items[0]

('fireball', 1)

In [217]:
#notice it is a tuple, so we need to sort based on the second number (the frequency)
sorted_items=sorted(items, key=lambda x: x[1], reverse=True)
sorted_items[:10]

[('to', 20),
 ('in', 14),
 ('and', 11),
 ('the', 11),
 ('for', 8),
 ('a', 7),
 ('says', 6),
 ('on', 6),
 ('is', 6),
 ('Dorian', 5)]

In [218]:
from sklearn.feature_extraction import stop_words
stop_words = set(stop_words.ENGLISH_STOP_WORDS)

In [219]:
#manual way
filtered_items = []
for item in sorted_items:
    if item[0] not in stop_words:
        filtered_items.append(item)

In [220]:
filtered_items[:10]

[('says', 6),
 ('Dorian', 5),
 ('Hurricane', 5),
 ('Trump', 5),
 ('gun', 4),
 ('Congress', 3),
 ('Donald', 3),
 ('California', 3),
 ("Here's", 3),
 ('Brexit', 3)]

In [221]:
#sneak peek, using list comprehension
[item for item in sorted_items if item[0] not in stop_words][:10]

[('says', 6),
 ('Dorian', 5),
 ('Hurricane', 5),
 ('Trump', 5),
 ('gun', 4),
 ('Congress', 3),
 ('Donald', 3),
 ('California', 3),
 ("Here's", 3),
 ('Brexit', 3)]

Use comprehensions and reduce to answer the question below

In [222]:
msg="""is it possible to star emails from this message for example
if there is an email such as doctor.fun@python.org replace it with stars,
another one is fun.ction@dawg.edu"""

In [223]:
' '.join(
    [
        word if '@' not in word else '*' * len(word)
        for word in msg.split()
    ]
)

'is it possible to star emails from this message for example if there is an email such as ********************* replace it with stars, another one is ******************'

What is the longest word in msg [answer is "doctor.fun@python.org"]

In [224]:
# using filtering
max_length = max([len(word) for word in msg.split()])
[word for word in msg.split() if len(word) == max_length]

['doctor.fun@python.org']

In [225]:
# using reduce
def longer(word1, word2):
    longer_word = word1 if len(word1) > len(word2) else word2
    return longer_word

# example
longer("java", "python")

'python'

In [226]:
from functools import reduce

reduce(longer, msg.split())

'doctor.fun@python.org'

In [227]:
# or using lambda

reduce(lambda w1, w2: w1 if len(w1) > len(w2) else w2, msg.split())

'doctor.fun@python.org'

What is the longest word in msg that is not an email [answer is "possible"]

In [ ]:
# using filtering
max_length = max([len(word) for word in msg.split() if '@' not in word])
[word for word in msg.split() if len(word) == max_length]

In [230]:
reduce(longer,
       [word for word in msg.split() if '@' not in word]
      )

'possible'

In [64]:
import csv

In [65]:
data = []
with open('data/EC10.csv') as csvfile:
    csvreader = csv.reader(csvfile, delimiter=',')
    for row in csvreader:
        data.append(row[1])
data = data[1:] #exclude header

FileNotFoundError: [Errno 2] No such file or directory: 'data/EC10.csv'

In [66]:
len(data)

0

In [67]:
print(data[0][:1000])

IndexError: list index out of range

The list `data` contains 10 entries. Each entry is the full-text of  a company's [earning call](https://en.wikipedia.org/wiki/Earnings_call)

The company's ticker and stock exchange are listed in the first few lines in each earning call. Using regular expressions, extract the tickers and stock exchange symbols for all the companies. 

Note that an earning call may contain tickers of other companies. We want to exclude them. In other words, we want to first ticker / stock exchange listed in each earning call

In [68]:
import re
stock_ticker = re.compile('[(]([A-Z]+):([A-Z]+)[)]')
stock_ticker.findall(data[0]) # just to test

IndexError: list index out of range

In [69]:
def extract_stock_ticker(earning_call, extract='both'):
    stock_ticker = re.compile('[(]([A-Z]+):([A-Z]+)[)]')
    stock_ticker.search(data[0])
    res = stock_ticker.findall(earning_call) 
    if res:
        hit = res[0] #first hit
        if extract == 'both':
            return hit
        elif extract == 'ticker':
            return hit[1]
        elif extract == 'exchange':
            return hit[0]
    return None # no need here, but to add an emphasis

In [70]:
tickers_exchanges = [extract_stock_ticker(x) for x in data]
tickers_exchanges


[]

In [71]:
tickers = [extract_stock_ticker(x, extract='ticker') for x in data]
tickers

[]

In [72]:
exchanges = [extract_stock_ticker(x, extract='exchange') for x in data]
exchanges

[]

How about extracting all numbers that start with `$`

In [73]:
money = re.compile(r'\$\d+[.,\d]+')
money.findall(data[0]) # just to test

IndexError: list index out of range

The numbers are meaningless so let us put some context

In [74]:
money = re.compile(r'(\w+\s)(\$\d+[.,\d]+)(\s\w+)')
money.findall(data[0]) # just to test

IndexError: list index out of range

Apply to all earning calls

In [75]:
money_extract = [ money.findall(x) for x in data ]

In [76]:
money_extract

[]

# Homework 1 Questions


# Instructions

Your submission will be tested with the code tester. It is important to follow these instructions to ensure your work tests properly.

- Do not change the content of the cells under __SETUP__ and __TESTS__
- Work only in the __YOUR WORK__ area
- Rename the notebook with your group at the end (subsitute XX with your group number).
- Assign the results of each numbered question to the appropriate test variable. For example, the answer of `1.` should be assigned to `test_1`
- Rounding: use the supplied function `hround` to round decimal numbers when instructed. It's important to use this function because there are [multiple ways to round numbers in Python](https://www.knowledgehut.com/blog/programming/python-rounding-numbers) and they may not result in the same value that the tester is testing against.
- Ensure your run the cells under __SETUP__ before you run your work
- Before you submit your work, ensure you clean up your notebook. Your notebook has to run without an error in order to be tested. The easiest way to ensure is to `Kernel->Restart & Run All`
- Answers are provided below for your convenience
- You will need to write a program to calculate the answers. Setting the answers to be their correct values without solving them is considered *hardcoding* and will result in zero grade for the assignment as well as a potential academic honesty violation.
- You can also test your submission using [the online code tester](http://open-data.stream:9100)


In [2]:
# DO NOT EDIT OR CHANGE THE CONTENT OF THIS CELL
scenario = 0

In [3]:
#imports
import random

In [4]:
def hround(number):
    return round(number, 2 - scenario)

In [8]:
test_1=test_2=test_3=test_4=test_5=test_6=test_7=test_8=test_9=0.0

## Part 1: Supertankers

*this question is inspired by Scott Page's [The Model Thinker](https://www.goodreads.com/book/show/39088592-the-model-thinker) Chapter 3*

The cost of building a supertanker depends primarily on its surface area, which determines the amount of steel used. The amount of revenue a supertanker generates depends on its volume. Computing the ratio of volume to surface area reveals a linear gain in profitability from increasing size.

To simplify the calculations below, assume a rectangular prism supertanker. 

\begin{align}
Surface area &= 2 \times (Beam \times Height + Height \times Draft + Beam \times Draft) \\
Volume &= Beam \times Height \times Draft
\end{align}

The T2 oil tanker used during World War II measured 500 feet long, 25 feet deep, and 50 feet wide.
1. What is the surface area of the T2 tanker?
2. What is the volume of the T2 tanker?
3. What is the ratio of volume to surface of the T2 tanker? Round the value to two decimals using the function `hround`. 

## Part 1: Answer

1. What is the surface area of the T2 tanker?
2. What is the volume of the T2 tanker?
3. What is the ratio of volume to surface of the T2 tanker? Round the value to two decimals using the function hround.

In [10]:
beam = 500
height = 25
draft = 50

test_1 = 2*(beam*height + beam*draft + height*draft)
print(test_1)

test_2 = beam*height*draft
print(test_2)

test_3 = hround(test_2/test_1)
print(test_3)

77500
625000
8.06


## Part 2: Flipping coins
We toss a fair coin five times and record the sequence of outcomes (H or T). 
4. How many possible sequences are there?
5. What is the chance that we end up with four heads or more? Round the number to two decimals.

Hints:
- If you took a college course in probability or statistics you may already know there are  formulas to calculate this probability. You can google "permutation and combination in python" as a start.
- Another way is to model the underlying distribution of outcomes as a [Binomial distribution](https://en.wikipedia.org/wiki/Binomial_distribution).
- However, you don't need to learn the underlying math. A third approach is computational. Generate all possible sequences of length five, then count how many of them have more than three heads. Below is a starter code that generates sequences of two flips

```python
outcomes = []
for f1 in ['H', 'T']:
    for f2 in ['H', 'T']:
        outcomes.append(f1+f2)
outcomes```

## Part 2: Answer

In [11]:
test_4 = 2**5
print(test_4)

import math
test_5 = (math.factorial(5)/math.factorial(4) + math.factorial(5)/math.factorial(5)) / test_4
print(test_5)

32
0.1875


## Part 3: Poker

OK, flipping coin is boring, how about playing poker? You don't need to know how to play poker to answer this question. Just read [the Wikipedia entry](https://en.wikipedia.org/wiki/List_of_poker_hands) about the different kinds of hands. To simplify the question, we will assume no Joker in our deck of cards.

We represent a poker hand as a list of tuples. A hand has five cards and hence five tuples. Each tuple has two components, the first is the rank and the second is the suit. Rank is one value from `['2','3','4','5','6','7','8','9','10','J','Q', 'K', 'A']` and suit is one value from `['♠', '♥', '♣', '♦']`

The function `draw_hand()` is provided below for your convinience. It returns a random hand of five from a deck of cards. For example, `[('2', '♥'), ('8', '♦'), ('2', '♣'), ('J', '♠'), ('6', '♣')]`. You can use the function to generate hands to test your work.

In this exercise, you will write two functions to test if a hand is a straight hand and a straight flush hand:
- A straight hand is a hand in which the cards' ranks are in order
- A straight flush hand is a straight hand in which the cards are from one suit

Write a function `straight_hand` that takes a hand as a parameter and return a boolean value.

6. Test the function `[('2', '♥'), ('8', '♦'), ('2', '♣'), ('J', '♠'), ('6', '♣')]`
7. Test the function using `[('9', '♥'), ('8', '♦'), ('10', '♣'), ('Q', '♠'), ('J', '♣')]`

Write a function `straight_flush_hand` that takes a hand as a parameter and return a boolean value.

8. Test the function using `[('9', '♥'), ('8', '♦'), ('10', '♣'), ('Q', '♠'), ('J', '♣')]`
9. Test the function using `[('9', '♥'), ('8', '♥'), ('10', '♥'), ('Q', '♥'), ('J', '♥')]`

**BONUS**: In a random draw, what is the chance of getting a straight hand? What is the chance of getting a straight flush hand?

## Part 3: Answer


In [12]:
def draw_hand():
    ranks = ['2','3','4','5','6','7','8','9','10','J','Q', 'K', 'A']
    suits = ['♠', '♥', '♣', '♦']
    deck = [(r, s) for s in suits for r in ranks ]
    hand = random.sample(deck, 5)
    return hand

In [13]:
def straight_hand(hand):
    values = []
    face_map = {'J': 11, 'Q': 12, 'K': 13, 'A': 14}
    for h in hand:
        try:
            values.append(int(h[0]))
        except:
            values.append(face_map[h[0]])    
    values = sorted(values)
    return (values == list(range(values[0], values[0]+5))
            or values == [2, 3, 4, 5, 14])

In [14]:
def flush_hand(hand):
    suits = [h[1] for h in hand]
    return all(s == suits[0] for s in suits)

In [15]:
def straight_flush_hand(hand):
    return straight_hand(hand) == True and flush_hand(hand) == True

In [16]:
test_6 = straight_hand([('2', '♥'), ('8', '♦'), ('2', '♣'), ('J', '♠'), ('6', '♣')])
print(test_6)

test_7 = straight_hand([('9', '♥'), ('8', '♦'), ('10', '♣'), ('Q', '♠'), ('J', '♣')])
print(test_7)

test_8 = straight_flush_hand([('9', '♥'), ('8', '♦'), ('10', '♣'), ('Q', '♠'), ('J', '♣')])
print(test_8)

test_9 = straight_flush_hand([('9', '♥'), ('8', '♥'), ('10', '♥'), ('Q', '♥'), ('J', '♥')])
print(test_9)

False
True
False
True


## Homework Answers


In [17]:
### TEST 1
test_1

77500

In [18]:
## TEST 2
test_2

625000

In [19]:
## TEST 3
test_3

8.06

In [20]:
## TEST 4
test_4

32

In [21]:
## TEST 5
test_5

0.1875

In [22]:
## TEST 6
test_6

False

In [23]:
## TEST 7
test_7

True

In [24]:
## TEST 8
test_8

False

In [25]:
## TEST 9
test_9

True

## Exam Review